README.md
[![Crates.io](https://img.shields.io/crates/v/overloadf.svg)](https://crates.io/crates/overloadf)
[![License](https://img.shields.io/crates/l/overloadf)](LICENSE-MIT)
[![Build Status](https://travis-ci.org/zenixls2/overloadf.svg?branch=master)](https://travis-ci.org/zenixls2/overloadf)

# overloadf version - 0.1.7

## Overloadf

** Let function overloading possible in rust **

With a single attribute on top of the function, you can overload the function with different
parameters. Current implementation still has some flaws and todo items, so use at your own
risk.

This library is based on some unstable features.
To use this library, please put the following lines in crate root and the beginning of test
files:
```rust
#![feature(fn_traits, unboxed_closures)]
```

There are some features that cannot be achieved until now:
- unsafe function overloading
- const function overloading
- different privacy setting on function overloading (will pickup the privacy setting in first
function and apply to all)
- function overloading inside traits (for limited cases)

### Examples:
simple one:
```rust
#![feature(fn_traits, unboxed_closures)]

use overloadf::*;

#[overload]
pub fn xdd() -> i32 {
    5_i32
}

#[overload]
pub fn xdd(number: i32) -> i32 {
    number * 3
}

#[overload]
pub fn xdd(number: u8) {
    println!("{}", number);
}

#[overload]
pub unsafe fn xdd(number: &u64) -> u64 {
    let n = number as *const u64;
    *n * 4
}

assert_eq!(xdd(3_i32), 9_i32);
let c: &u64 = &6_u64;
assert_eq!(xdd(c), 24_u64); // unsafe function is not supported.
assert_eq!(xdd(), 5_i32);
```

with generic and custom type:
```rust
#![feature(fn_traits, unboxed_closures)]

use overloadf::*;
use std::ops::MulAssign;
use std::fmt::Debug;
#[overload]
pub fn xdd<T: Copy + Debug + MulAssign<i32>>(mut number: T) -> T
where
    T: PartialEq,
{
    println!("number {:?}", number);
    number *= 3_i32;
    number
}

struct ABC;
impl Debug for ABC {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "xdddd")
    }
}

#[overload]
pub fn xdd(number: ABC) -> () {
    println!("number {:?}", number);
}
let aa = 123;
assert_eq!(xdd(aa), 369);
assert_eq!(xdd(ABC), ());
```
async / await:
```rust
#![feature(fn_traits, unboxed_closures)]
use overloadf::*;

#[overload]
pub async fn xdd(number: i32) -> i32 {
    number + 3
}
#[overload]
pub async fn xdd(number: i64) -> i64 {
    number + 4
}
assert_eq!(futures::executor::block_on(xdd(3_i32)), 6);
assert_eq!(futures::executor::block_on(xdd(3_i64)), 7);
```

type conflict might happen if generic overlaps with the definition of implemented types:
```rust,compile_fail
#![feature(fn_traits, unboxed_closures)]

use overloadf::*;
use std::ops::Mul;
use std::fmt::Debug;
#[overload]
pub fn xdd(number: i32) -> i32 {
    number * 2
}
#[overload]
pub fn xdd<T: Copy + Debug + Mul<i32>>(number: T) -> T {
    number * 3_i32
}
```

for trait methods (notice that trait for overload must inherit Sized):
```rust
#![feature(fn_traits, unboxed_closures)]
use overloadf::*;
#[overload]
trait Xdd: Sized {
    fn new(input: i32) -> Self;
    fn new(input :u32) -> Self;
}
struct Haha {
    a: u32,
    b: i32,
}
#[overload]
impl Xdd for Haha {
    fn new(b: i32) -> Self {
        Self {
            a: 1,
            b,
        }
    }
    fn new(a: u32) -> Self {
        Self {
            a,
            b: 2,
        }
    }
}
let haha = Haha::new(12_i32);
assert_eq!(haha.a, 1_u32);
assert_eq!(haha.b, 12_i32);
let haha = Haha::new(9_u32);
assert_eq!(haha.a, 9_u32);
assert_eq!(haha.b, 2_i32);
```

dynamic trait object implementation:
```rust
#![feature(fn_traits, unboxed_closures)]
use overloadf::*;
trait Xdd: 'static {}

#[overload]
impl dyn Xdd {
    fn abc(&self, c: i32) -> i32 {
        c
    }
    fn abc(&self, c: u32) -> u32 {
        c
    }
}

impl<T: 'static + ?Sized> Xdd for T {}
let obj = (&3_i32 as &(dyn Xdd));
// here we lost the syntax sugar of calling obj.abc(c) directly
assert_eq!(Xdd::abc(obj, 3_i32), 3_i32);
assert_eq!(Xdd::abc(obj, 3_u32), 3_u32);
```

trait with generics:
```rust
#![feature(fn_traits, unboxed_closures)]
use overloadf::*;
#[overload]
trait Xdd<T: Sized>: Sized {
    type J: Into<i32>;
    fn new(input: i32) -> T where T: Debug;
    // trait function with default implementation
    fn new(input: u32) -> String {
        "please use new(input1, input2) instead".to_string()
    }
    fn new<I: Into<u32>>(input1: I, input2: Self::J) -> T where T: Debug;
}
#[derive(Debug)]
struct Haha {
    a: u32,
    b: i32,
}
#[overload]
impl Xdd<Haha> for Haha {
    type J = i32;
    fn new(b: i32) -> Haha {
        Self {
            a: 1,
            b,
        }
    }
    fn new<I: Into<u32>>(a: I, b: Self::J) -> Haha {
        Self {
            a: a.into(),
            b,
        }
    }
}
let haha = Haha::new(12_i32);
assert_eq!(haha.a, 1_u32);
assert_eq!(haha.b, 12_i32);
let haha = Haha::new(9_u32, 2_i32);
assert_eq!(haha.a, 9_u32);
assert_eq!(haha.b, 2_i32);
assert_eq!(Haha::new(3_u32), "please use new(input1, input2) instead".to_string());
```

non-trait impl:
```rust
#![feature(fn_traits, unboxed_closures)]
use overloadf::*;
#[derive(Debug)]
pub struct Haha {
    a: u32,
    b: i32,
}
#[overload]
impl Haha {
    pub fn new(b: i32) -> Self {
        Self {
            a: 1,
            b,
        }
    }
    pub fn new(a: u32) -> Self {
        Self {
            a,
            b: 2,
        }
    }
    // self in function parameter will be converted to associated type
    // that means no more syntax sugar like haha.normal()
    pub fn normal(&self) -> String {
        format!("{:?}", self)
    }
    pub fn normal(&self, prefix: &str) -> String {
        format!("{} {:?}", prefix, self)
    }
    // function without overloading is not influenced
    pub fn display(&self) -> String {
        format!("{:?}", self)
    }
}
let haha = Haha::new(12_i32);
assert_eq!(haha.a, 1_u32);
assert_eq!(haha.b, 12_i32);
let haha = Haha::new(9_u32);
assert_eq!(haha.a, 9_u32);
assert_eq!(haha.b, 2_i32);
assert_eq!(Haha::normal(&haha), "Haha { a: 9, b: 2 }");
assert_eq!(Haha::normal(&haha, "abc"), "abc Haha { a: 9, b: 2 }");
assert_eq!(haha.display(), "Haha { a: 9, b: 2 }");
```

default attribute:

in functions(not yet for traits), now you could decorate parameters with default values.
the reason why we don't support `parameter = value` syntax directly is that in derived
`TokenStream`, values inside will go through compiler first for syntax check.
The default value syntax will be treated as a compile error and forbid us from parsing
and generating valid Tokens.

example:
```rust
#![feature(fn_traits, unboxed_closures)]
use overloadf::*;
#[overload]
fn xdd(#[default(= 5_i32)] a: i32, #[default(= 32_u64)] b: u64) -> u64 { b - (a as u64) }
assert_eq!(xdd(3_i32), 29_u64);
assert_eq!(xdd(4_i32, 4_u64), 0);
assert_eq!(xdd(), 27_u64);
// compile error: expect i32 found u64
// assert_eq!(xdd(6_u64), 1_u64);
```

```rust
#![feature(fn_traits, unboxed_closures)]
use overloadf::*;
#[overload]
fn xdd(#[default(= 5_i32)] a: i32, b: u8, #[default(= 32_u64)] c: u64) -> u64 { c + (b as u64) - (a as u64) }
assert_eq!(xdd(4_i32, 7_u8), 35_u64);
assert_eq!(xdd(3_u8), 30_u64);
```

### License

Licensed under

* MIT license ([LICENSE-MIT](LICENSE-MIT) or https://opensource.org/licences/MIT)

### Contribution

Unless you explicitly state otherwise, any contribution intentionally
submitted for inclusion in the work by you, shall be licensed as above,
without any additional terms or conditions.


CHANGELOG.md
0.1.7
=====

replace `std` with `core` in generated code
fix dyn Trait implementation

0.1.6
=====

fix functions with empty input
add default parameters support for normal functions

0.1.5
=====

fix return type = () in Output type binding
fix derive version

0.1.4
=====

implement trait default method overloading (i.e., trait methods with default block)

0.1.3
=====

correctly handle `self` and `Self` in trait methods.
Handle associated types in parameters.

0.1.2
=====

support where clause by correctly making use of `syn::Generics`

0.1.1
=====

start to support trait overloading

0.1.0
=====

add function overloading


overloadf_derive/src/fn_struct.rs
use quote::{ToTokens, TokenStreamExt};
use std::iter::FromIterator;
use syn::parse::{discouraged::Speculative, Parse, ParseStream, Result};
use syn::punctuated::{Pair, Punctuated};
trait FilterAttrs<'a> {
    type Ret: Iterator<Item = &'a syn::Attribute>;
    fn outer(self) -> Self::Ret;
    fn inner(self) -> Self::Ret;
}

impl<'a, T> FilterAttrs<'a> for T
where
    T: IntoIterator<Item = &'a syn::Attribute>,
{
    type Ret = std::iter::Filter<T::IntoIter, fn(&&syn::Attribute) -> bool>;

    fn outer(self) -> Self::Ret {
        fn is_outer(attr: &&syn::Attribute) -> bool {
            match attr.style {
                syn::AttrStyle::Outer => true,
                _ => false,
            }
        }
        self.into_iter().filter(is_outer)
    }

    fn inner(self) -> Self::Ret {
        fn is_inner(attr: &&syn::Attribute) -> bool {
            match attr.style {
                syn::AttrStyle::Inner(_) => true,
                _ => false,
            }
        }
        self.into_iter().filter(is_inner)
    }
}

#[derive(Debug, Eq, PartialEq, Hash, Clone)]
pub struct Assign {
    pub eq_token: syn::token::Eq,
    pub right: Box<syn::Expr>,
}

#[derive(Debug, Eq, PartialEq, Hash, Clone)]
pub struct PatType {
    pub attrs: Vec<syn::Attribute>,
    pub pat: Box<syn::Pat>,
    pub colon_token: Token![:],
    pub ty: Box<syn::Type>,
    pub assign: Option<Assign>,
}

#[derive(Debug, Eq, PartialEq, Hash, Clone)]
pub enum FnArg {
    Receiver(syn::Receiver),
    Typed(PatType),
}

#[derive(Debug, Eq, PartialEq, Hash, Clone)]
pub struct Signature {
    pub constness: Option<Token![const]>,
    pub asyncness: Option<Token![async]>,
    pub unsafety: Option<Token![unsafe]>,
    pub abi: Option<syn::Abi>,
    pub fn_token: Token![fn],
    pub ident: syn::Ident,
    pub generics: syn::Generics,
    pub paren_token: syn::token::Paren,
    pub inputs: Punctuated<FnArg, Token![,]>,
    pub variadic: Option<syn::Variadic>,
    pub output: syn::ReturnType,
}

// the ItemFn that supports default arguments
#[derive(Debug, Eq, PartialEq, Hash, Clone)]
pub struct ItemFn {
    pub attrs: Vec<syn::Attribute>,
    pub vis: syn::Visibility,
    pub sig: Signature,
    pub block: Box<syn::Block>,
}

impl Parse for Assign {
    fn parse(input: ParseStream) -> Result<Self> {
        let eq_token: Token![=] = input.parse()?;
        let right: Box<syn::Expr> = input.parse()?;
        Ok(Self { eq_token, right })
    }
}

fn fn_arg_typed(input: ParseStream) -> Result<PatType> {
    // Hack to parse pre-2018 syntax in
    // test/ui/rfc-2565-param-attrs/param-attrs-pretty.rs
    // because the rest of the test case is valuable.
    if input.peek(syn::Ident) && input.peek2(Token![<]) {
        let span = input.fork().parse::<syn::Ident>()?.span();
        let mut pt = PatType {
            attrs: Vec::new(),
            pat: Box::new(syn::Pat::Wild(syn::PatWild {
                attrs: Vec::new(),
                underscore_token: Token![_](span),
            })),
            colon_token: Token![:](span),
            ty: input.parse()?,
            assign: None,
        };
        if input.peek(Token![=]) {
            pt.assign = Some(input.parse()?);
        }
        return Ok(pt);
    }

    let pat: Box<syn::Pat> = input.parse()?;
    let colon_token: syn::token::Colon = input.parse()?;
    let ty: syn::Type = match input.parse::<Option<Token![...]>>()? {
        Some(dot3) => syn::Type::Verbatim(variadic_to_tokens(&dot3)),
        None => input.parse()?,
    };
    println!("here {:?}", input);
    let assign: Option<Assign> = if input.peek(Token![=]) {
        Some(input.parse()?)
    } else {
        None
    };
    Ok(PatType {
        attrs: Vec::new(),
        pat,
        colon_token,
        ty: Box::new(ty),
        assign,
    })
}

impl Parse for FnArg {
    fn parse(input: ParseStream) -> Result<Self> {
        let attrs = input.call(syn::Attribute::parse_outer)?;

        let ahead = input.fork();
        if let Ok(mut receiver) = ahead.parse::<syn::Receiver>() {
            if !ahead.peek(Token![:]) {
                input.advance_to(&ahead);
                receiver.attrs = attrs;
                return Ok(FnArg::Receiver(receiver));
            }
        }

        let mut typed = input.call(fn_arg_typed)?;
        typed.attrs = attrs;
        Ok(FnArg::Typed(typed))
    }
}

fn pop_variadic(args: &mut Punctuated<FnArg, Token![,]>) -> Option<syn::Variadic> {
    let trailing_punct = args.trailing_punct();

    let last = match args.last_mut()? {
        FnArg::Typed(last) => last,
        _ => return None,
    };

    let ty = match last.ty.as_ref() {
        syn::Type::Verbatim(ty) => ty,
        _ => return None,
    };

    let mut variadic = syn::Variadic {
        attrs: Vec::new(),
        dots: syn::parse2(ty.clone()).ok()?,
    };

    if let syn::Pat::Verbatim(pat) = last.pat.as_ref() {
        if pat.to_string() == "..." && !trailing_punct {
            variadic.attrs = std::mem::replace(&mut last.attrs, Vec::new());
            args.pop();
        }
    }

    Some(variadic)
}

fn parse_fn_args(input: ParseStream) -> Result<Punctuated<FnArg, Token![,]>> {
    let mut args = Punctuated::new();
    let mut has_receiver = false;

    while !input.is_empty() {
        let attrs = input.call(syn::Attribute::parse_outer)?;

        let arg = if let Some(dots) = input.parse::<Option<Token![...]>>()? {
            FnArg::Typed(PatType {
                attrs,
                pat: Box::new(syn::Pat::Verbatim(variadic_to_tokens(&dots))),
                colon_token: Token![:](dots.spans[0]),
                ty: Box::new(syn::Type::Verbatim(variadic_to_tokens(&dots))),
                assign: None,
            })
        } else {
            let mut arg: FnArg = input.parse()?;
            match &mut arg {
                FnArg::Receiver(receiver) if has_receiver => {
                    return Err(syn::Error::new(
                        receiver.self_token.span,
                        "unexpected second method receiver",
                    ));
                }
                FnArg::Receiver(receiver) if !args.is_empty() => {
                    return Err(syn::Error::new(
                        receiver.self_token.span,
                        "unexpected method receiver",
                    ));
                }
                FnArg::Receiver(receiver) => {
                    has_receiver = true;
                    receiver.attrs = attrs;
                }
                FnArg::Typed(arg) => arg.attrs = attrs,
            }
            arg
        };
        args.push_value(arg);

        if input.is_empty() {
            break;
        }

        let comma: Token![,] = input.parse()?;
        args.push_punct(comma);
    }

    Ok(args)
}

fn attrs(outer: Vec<syn::Attribute>, inner: Vec<syn::Attribute>) -> Vec<syn::Attribute> {
    let mut attrs = outer;
    attrs.extend(inner);
    attrs
}

impl Parse for ItemFn {
    fn parse(input: ParseStream) -> Result<Self> {
        println!("input {:?}", input);
        let outer_attrs = input.call(syn::Attribute::parse_outer)?;
        let vis: syn::Visibility = input.parse()?;
        let constness: Option<Token![const]> = input.parse()?;
        let asyncness: Option<Token![async]> = input.parse()?;
        let unsafety: Option<Token![unsafe]> = input.parse()?;
        let abi: Option<syn::Abi> = input.parse()?;
        let fn_token: Token![fn] = input.parse()?;
        let ident: syn::Ident = input.parse()?;
        let generics: syn::Generics = input.parse()?;

        let content;
        let paren_token = parenthesized!(content in input);
        println!("content {:?}", content);
        let mut inputs = parse_fn_args(&content)?;
        let variadic = pop_variadic(&mut inputs);

        let output: syn::ReturnType = input.parse()?;
        let where_clause: Option<syn::WhereClause> = input.parse()?;

        let content;
        let brace_token = braced!(content in input);
        let inner_attrs = content.call(syn::Attribute::parse_inner)?;
        let stmts = content.call(syn::Block::parse_within)?;

        Ok(ItemFn {
            attrs: attrs(outer_attrs, inner_attrs),
            vis,
            sig: Signature {
                constness,
                asyncness,
                unsafety,
                abi,
                fn_token,
                ident,
                paren_token,
                inputs,
                output,
                variadic,
                generics: syn::Generics {
                    where_clause,
                    ..generics
                },
            },
            block: Box::new(syn::Block { brace_token, stmts }),
        })
    }
}

impl ToTokens for Assign {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        self.eq_token.to_tokens(tokens);
        self.right.to_tokens(tokens);
    }
}

impl ToTokens for PatType {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        tokens.append_all(self.attrs.outer());
        self.pat.to_tokens(tokens);
        self.colon_token.to_tokens(tokens);
        self.ty.to_tokens(tokens);
        self.assign.to_tokens(tokens);
    }
}

impl ToTokens for FnArg {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        match &*self {
            FnArg::Receiver(x) => x.to_tokens(tokens),
            FnArg::Typed(x) => x.to_tokens(tokens),
        }
    }
}

fn maybe_variadic_to_tokens(arg: &FnArg, tokens: &mut proc_macro2::TokenStream) -> bool {
    let arg = match arg {
        FnArg::Typed(arg) => arg,
        FnArg::Receiver(receiver) => {
            receiver.to_tokens(tokens);
            return false;
        }
    };

    match arg.ty.as_ref() {
        syn::Type::Verbatim(ty) if ty.to_string() == "..." => {
            match arg.pat.as_ref() {
                syn::Pat::Verbatim(pat) if pat.to_string() == "..." => {
                    tokens.append_all(arg.attrs.outer());
                    pat.to_tokens(tokens);
                }
                _ => arg.to_tokens(tokens),
            }
            true
        }
        _ => {
            arg.to_tokens(tokens);
            false
        }
    }
}

fn variadic_to_tokens(dots: &Token![...]) -> proc_macro2::TokenStream {
    proc_macro2::TokenStream::from_iter(vec![
        proc_macro2::TokenTree::Punct({
            let mut dot = proc_macro2::Punct::new('.', proc_macro2::Spacing::Joint);
            dot.set_span(dots.spans[0]);
            dot
        }),
        proc_macro2::TokenTree::Punct({
            let mut dot = proc_macro2::Punct::new('.', proc_macro2::Spacing::Joint);
            dot.set_span(dots.spans[1]);
            dot
        }),
        proc_macro2::TokenTree::Punct({
            let mut dot = proc_macro2::Punct::new('.', proc_macro2::Spacing::Joint);
            dot.set_span(dots.spans[2]);
            dot
        }),
    ])
}

impl ToTokens for Signature {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        self.constness.to_tokens(tokens);
        self.asyncness.to_tokens(tokens);
        self.unsafety.to_tokens(tokens);
        self.abi.to_tokens(tokens);
        self.fn_token.to_tokens(tokens);
        self.ident.to_tokens(tokens);
        self.generics.to_tokens(tokens);
        self.paren_token.surround(tokens, |tokens| {
            let mut last_is_variadic = false;
            for input in self.inputs.pairs() {
                match input {
                    Pair::Punctuated(input, comma) => {
                        maybe_variadic_to_tokens(input, tokens);
                        comma.to_tokens(tokens);
                    }
                    Pair::End(input) => {
                        last_is_variadic = maybe_variadic_to_tokens(input, tokens);
                    }
                }
            }
            if self.variadic.is_some() && !last_is_variadic {
                if !self.inputs.empty_or_trailing() {
                    <Token![,]>::default().to_tokens(tokens);
                }
                self.variadic.to_tokens(tokens);
            }
        });
        self.output.to_tokens(tokens);
        self.generics.where_clause.to_tokens(tokens);
    }
}

impl ToTokens for ItemFn {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        tokens.append_all(self.attrs.outer());
        self.vis.to_tokens(tokens);
        self.sig.to_tokens(tokens);
        self.block.brace_token.surround(tokens, |tokens| {
            tokens.append_all(self.attrs.inner());
            tokens.append_all(&self.block.stmts);
        });
    }
}


overloadf_derive/src/input_iter.rs
use crate::fn_struct;

pub struct InputIter {
    pub index: usize,
    pub input_types: Vec<syn::Type>,
    pub param_assign: Vec<syn::Pat>,
    pub default_values: Vec<Option<fn_struct::Assign>>,
    pub end: bool,
}

impl InputIter {
    pub fn new(a: Vec<syn::Type>, b: Vec<syn::Pat>, c: Vec<Option<fn_struct::Assign>>) -> Self {
        Self {
            index: a.len(),
            input_types: a,
            param_assign: b,
            default_values: c,
            end: false,
        }
    }
}

impl Iterator for InputIter {
    type Item = (Vec<syn::Type>, Vec<syn::Pat>, Vec<proc_macro2::TokenStream>);
    fn next(&mut self) -> Option<Self::Item> {
        let mut inputs = vec![];
        let mut params = vec![];
        let mut defaults = vec![];
        if self.end {
            return None;
        }
        let mut i = 0;
        loop {
            if i < self.index {
                inputs.push(self.input_types[i].clone());
                params.push(self.param_assign[i].clone());
                i += 1;
            } else {
                break;
            }
        }
        i = self.index;
        loop {
            if i < self.default_values.len() {
                if let Some(default) = &self.default_values[i] {
                    let p = &self.param_assign[i];
                    defaults.push(quote!(let #p #default;));
                } else {
                    inputs.push(self.input_types[i].clone());
                    params.push(self.param_assign[i].clone());
                }
                i += 1;
            } else {
                break;
            }
        }
        loop {
            if self.index > 0 {
                self.index -= 1;
                if self.default_values[self.index].is_some() {
                    break;
                }
            } else {
                self.end = true;
                break;
            }
        }
        Some((inputs, params, defaults))
    }
}


src/lib.rs
#![feature(fn_traits, unboxed_closures, const_extern_fn)]

//! # Overloadf
//!
//! ** Let function overloading possible in rust **
//!
//! With a single attribute on top of the function, you can overload the function with different
//! parameters. Current implementation still has some flaws and todo items, so use at your own
//! risk.
//!
//! This library is based on some unstable features.
//! To use this library, please put the following lines in crate root and the beginning of test
//! files:
//! ```rust,no_run
//! #![feature(fn_traits, unboxed_closures)]
//! ```
//!
//! There are some features that cannot be achieved until now:
//! - unsafe function overloading
//! - const function overloading
//! - different privacy setting on function overloading (will pickup the privacy setting in first
//! function and apply to all)
//! - function overloading inside traits (for limited cases)
//!
//! ## Examples:
//! simple one:
//! ```rust
//! #![feature(fn_traits, unboxed_closures)]
//!
//! use overloadf::*;
//!
//! #[overload]
//! pub fn xdd() -> i32 {
//!     5_i32
//! }
//!
//! #[overload]
//! pub fn xdd(number: i32) -> i32 {
//!     number * 3
//! }
//!
//! #[overload]
//! pub fn xdd(number: u8) {
//!     println!("{}", number);
//! }
//!
//! #[overload]
//! pub unsafe fn xdd(number: &u64) -> u64 {
//!     let n = number as *const u64;
//!     *n * 4
//! }
//!
//! assert_eq!(xdd(3_i32), 9_i32);
//! let c: &u64 = &6_u64;
//! assert_eq!(xdd(c), 24_u64); // unsafe function is not supported.
//! assert_eq!(xdd(), 5_i32);
//! ```
//!
//! with generic and custom type:
//! ```rust
//! #![feature(fn_traits, unboxed_closures)]
//!
//! use overloadf::*;
//! use std::ops::MulAssign;
//! use std::fmt::Debug;
//! #[overload]
//! pub fn xdd<T: Copy + Debug + MulAssign<i32>>(mut number: T) -> T
//! where
//!     T: PartialEq,
//! {
//!     println!("number {:?}", number);
//!     number *= 3_i32;
//!     number
//! }
//!
//! struct ABC;
//! impl Debug for ABC {
//!     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
//!         write!(f, "xdddd")
//!     }
//! }
//!
//! #[overload]
//! pub fn xdd(number: ABC) -> () {
//!     println!("number {:?}", number);
//! }
//! let aa = 123;
//! assert_eq!(xdd(aa), 369);
//! assert_eq!(xdd(ABC), ());
//! ```
//! async / await:
//! ```rust
//! #![feature(fn_traits, unboxed_closures)]
//! use overloadf::*;
//!
//! #[overload]
//! pub async fn xdd(number: i32) -> i32 {
//!     number + 3
//! }
//! #[overload]
//! pub async fn xdd(number: i64) -> i64 {
//!     number + 4
//! }
//! assert_eq!(futures::executor::block_on(xdd(3_i32)), 6);
//! assert_eq!(futures::executor::block_on(xdd(3_i64)), 7);
//! ```
//!
//! type conflict might happen if generic overlaps with the definition of implemented types:
//! ```rust,compile_fail
//! #![feature(fn_traits, unboxed_closures)]
//!
//! use overloadf::*;
//! use std::ops::Mul;
//! use std::fmt::Debug;
//! #[overload]
//! pub fn xdd(number: i32) -> i32 {
//!     number * 2
//! }
//! #[overload]
//! pub fn xdd<T: Copy + Debug + Mul<i32>>(number: T) -> T {
//!     number * 3_i32
//! }
//! ```
//!
//! for trait methods (notice that trait for overload must inherit Sized):
//! ```rust
//! #![feature(fn_traits, unboxed_closures)]
//! use overloadf::*;
//! #[overload]
//! trait Xdd: Sized {
//!     fn new(input: i32) -> Self;
//!     fn new(input :u32) -> Self;
//! }
//! struct Haha {
//!     a: u32,
//!     b: i32,
//! }
//! #[overload]
//! impl Xdd for Haha {
//!     fn new(b: i32) -> Self {
//!         Self {
//!             a: 1,
//!             b,
//!         }
//!     }
//!     fn new(a: u32) -> Self {
//!         Self {
//!             a,
//!             b: 2,
//!         }
//!     }
//! }
//! let haha = Haha::new(12_i32);
//! assert_eq!(haha.a, 1_u32);
//! assert_eq!(haha.b, 12_i32);
//! let haha = Haha::new(9_u32);
//! assert_eq!(haha.a, 9_u32);
//! assert_eq!(haha.b, 2_i32);
//! ```
//!
//! dynamic trait object implementation:
//! ```rust
//! #![feature(fn_traits, unboxed_closures)]
//! use overloadf::*;
//! trait Xdd: 'static {}
//!
//! #[overload]
//! impl dyn Xdd {
//!     fn abc(&self, c: i32) -> i32 {
//!         c
//!     }
//!     fn abc(&self, c: u32) -> u32 {
//!         c
//!     }
//! }
//!
//! impl<T: 'static + ?Sized> Xdd for T {}
//! let obj = (&3_i32 as &(dyn Xdd));
//! // here we lost the syntax sugar of calling obj.abc(c) directly
//! assert_eq!(Xdd::abc(obj, 3_i32), 3_i32);
//! assert_eq!(Xdd::abc(obj, 3_u32), 3_u32);
//! ```
//!
//! trait with generics:
//! ```rust
//! #![feature(fn_traits, unboxed_closures)]
//! use overloadf::*;
//! #[overload]
//! trait Xdd<T: Sized>: Sized {
//!     type J: Into<i32>;
//!     fn new(input: i32) -> T where T: Debug;
//!     // trait function with default implementation
//!     fn new(input: u32) -> String {
//!         "please use new(input1, input2) instead".to_string()
//!     }
//!     fn new<I: Into<u32>>(input1: I, input2: Self::J) -> T where T: Debug;
//! }
//! #[derive(Debug)]
//! struct Haha {
//!     a: u32,
//!     b: i32,
//! }
//! #[overload]
//! impl Xdd<Haha> for Haha {
//!     type J = i32;
//!     fn new(b: i32) -> Haha {
//!         Self {
//!             a: 1,
//!             b,
//!         }
//!     }
//!     fn new<I: Into<u32>>(a: I, b: Self::J) -> Haha {
//!         Self {
//!             a: a.into(),
//!             b,
//!         }
//!     }
//! }
//! let haha = Haha::new(12_i32);
//! assert_eq!(haha.a, 1_u32);
//! assert_eq!(haha.b, 12_i32);
//! let haha = Haha::new(9_u32, 2_i32);
//! assert_eq!(haha.a, 9_u32);
//! assert_eq!(haha.b, 2_i32);
//! assert_eq!(Haha::new(3_u32), "please use new(input1, input2) instead".to_string());
//! ```
//!
//! non-trait impl:
//! ```rust
//! #![feature(fn_traits, unboxed_closures)]
//! use overloadf::*;
//! #[derive(Debug)]
//! pub struct Haha {
//!     a: u32,
//!     b: i32,
//! }
//! #[overload]
//! impl Haha {
//!     pub fn new(b: i32) -> Self {
//!         Self {
//!             a: 1,
//!             b,
//!         }
//!     }
//!     pub fn new(a: u32) -> Self {
//!         Self {
//!             a,
//!             b: 2,
//!         }
//!     }
//!     // self in function parameter will be converted to associated type
//!     // that means no more syntax sugar like haha.normal()
//!     pub fn normal(&self) -> String {
//!         format!("{:?}", self)
//!     }
//!     pub fn normal(&self, prefix: &str) -> String {
//!         format!("{} {:?}", prefix, self)
//!     }
//!     // function without overloading is not influenced
//!     pub fn display(&self) -> String {
//!         format!("{:?}", self)
//!     }
//! }
//! let haha = Haha::new(12_i32);
//! assert_eq!(haha.a, 1_u32);
//! assert_eq!(haha.b, 12_i32);
//! let haha = Haha::new(9_u32);
//! assert_eq!(haha.a, 9_u32);
//! assert_eq!(haha.b, 2_i32);
//! assert_eq!(Haha::normal(&haha), "Haha { a: 9, b: 2 }");
//! assert_eq!(Haha::normal(&haha, "abc"), "abc Haha { a: 9, b: 2 }");
//! assert_eq!(haha.display(), "Haha { a: 9, b: 2 }");
//! ```
//!
//! default attribute:
//!
//! in functions(not yet for traits), now you could decorate parameters with default values.
//! the reason why we don't support `parameter = value` syntax directly is that in derived
//! `TokenStream`, values inside will go through compiler first for syntax check.
//! The default value syntax will be treated as a compile error and forbid us from parsing
//! and generating valid Tokens.
//!
//! example:
//! ```rust
//! #![feature(fn_traits, unboxed_closures)]
//! use overloadf::*;
//! #[overload]
//! fn xdd(#[default(= 5_i32)] a: i32, #[default(= 32_u64)] b: u64) -> u64 { b - (a as u64) }
//! assert_eq!(xdd(3_i32), 29_u64);
//! assert_eq!(xdd(4_i32, 4_u64), 0);
//! assert_eq!(xdd(), 27_u64);
//! // compile error: expect i32 found u64
//! // assert_eq!(xdd(6_u64), 1_u64);
//! ```
//!
//! ```rust
//! #![feature(fn_traits, unboxed_closures)]
//! use overloadf::*;
//! #[overload]
//! fn xdd(#[default(= 5_i32)] a: i32, b: u8, #[default(= 32_u64)] c: u64) -> u64 { c + (b as u64) - (a as u64) }
//! assert_eq!(xdd(4_i32, 7_u8), 35_u64);
//! assert_eq!(xdd(3_u8), 30_u64);
//! ```

pub extern crate overloadf_derive;
pub use overloadf_derive::overload;
